File: src/lib/api/knowledge.ts
import { AppServerClient } from "@/lib/api/appServer";
import {
  METHOD_KNOWLEDGE_CONTEXT_RESOLVE,
  METHOD_KNOWLEDGE_CONTEXT_RUN_VALIDATE,
  METHOD_KNOWLEDGE_PACK_COMPILE,
  METHOD_KNOWLEDGE_PACK_DEFAULT_SET,
  METHOD_KNOWLEDGE_PACK_LIST,
  METHOD_KNOWLEDGE_PACK_READ,
  METHOD_KNOWLEDGE_PACK_STATUS_UPDATE,
  METHOD_KNOWLEDGE_SOURCE_IMPORT,
  type KnowledgeCompilePackResponse as AppServerKnowledgeCompilePackResponse,
  type KnowledgeContextResolutionResponse as AppServerKnowledgeContextResolutionResponse,
  type KnowledgeImportSourceResponse as AppServerKnowledgeImportSourceResponse,
  type KnowledgeListPacksResponse as AppServerKnowledgeListPacksResponse,
  type KnowledgeReadPackResponse as AppServerKnowledgeReadPackResponse,
  type KnowledgeSetDefaultPackResponse as AppServerKnowledgeSetDefaultPackResponse,
  type KnowledgeUpdatePackStatusResponse as AppServerKnowledgeUpdatePackStatusResponse,
  type KnowledgeValidateContextRunResponse as AppServerKnowledgeValidateContextRunResponse,
} from "../../../packages/app-server-client/src/protocol";

export type KnowledgeAppServerClient = Pick<AppServerClient, "request">;

export interface KnowledgeListPacksOptions {
  appServerClient?: KnowledgeAppServerClient;
}

async function requestKnowledgeAppServer<T>(
  method: string,
  params?: unknown,
  appServerClient: KnowledgeAppServerClient = new AppServerClient(),
): Promise<T> {
  const response = await appServerClient.request<T>(method, params);
  return response.result;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function assertKnowledgePackDetail(
  command: string,
  value: unknown,
): asserts value is KnowledgePackDetail {
  if (
    !isRecord(value) ||
    !isRecord(value.metadata) ||
    typeof value.rootPath !== "string"
  ) {
    throw new Error(`${command} did not return a knowledge pack detail`);
  }
}

function assertKnowledgeFileEntry(
  command: string,
  value: unknown,
  label: string,
): asserts value is KnowledgePackFileEntry {
  if (!isRecord(value) || typeof value.relativePath !== "string") {
    throw new Error(`${command} did not return ${label}`);
  }
}

function assertKnowledgeImportSourceResponse(
  command: string,
  value: unknown,
): asserts value is KnowledgeImportSourceResponse {
  if (!isRecord(value)) {
    throw new Error(`${command} did not return an import source result`);
  }
  assertKnowledgePackDetail(command, value.pack);
  assertKnowledgeFileEntry(command, value.source, "an imported source file");
}

function assertKnowledgeCompilePackResponse(
  command: string,
  value: unknown,
): asserts value is KnowledgeCompilePackResponse {
  if (
    !isRecord(value) ||
    typeof value.selectedSourceCount !== "number" ||
    !Array.isArray(value.warnings)
  ) {
    throw new Error(`${command} did not return a compile result`);
  }
  assertKnowledgePackDetail(command, value.pack);
  assertKnowledgeFileEntry(command, value.compiledView, "a compiled view file");
  assertKnowledgeFileEntry(command, value.run, "a compile run file");
}

function assertKnowledgeSetDefaultPackResponse(
  command: string,
  value: unknown,
): asserts value is KnowledgeSetDefaultPackResponse {
  if (
    !isRecord(value) ||
    typeof value.defaultPackName !== "string" ||
    typeof value.defaultMarkerPath !== "string"
  ) {
    throw new Error(`${command} did not return a default pack result`);
  }
}

function assertKnowledgeUpdatePackStatusResponse(
  command: string,
  value: unknown,
): asserts value is KnowledgeUpdatePackStatusResponse {
  if (
    !isRecord(value) ||
    typeof value.previousStatus !== "string" ||
    typeof value.clearedDefault !== "boolean"
  ) {
    throw new Error(`${command} did not return a status update result`);
  }
  assertKnowledgePackDetail(command, value.pack);
}

function assertKnowledgeContextResolution(
  command: string,
  value: unknown,
): asserts value is KnowledgeContextResolution {
  if (
    !isRecord(value) ||
    typeof value.packName !== "string" ||
    typeof value.status !== "string" ||
    typeof value.fencedContext !== "string" ||
    typeof value.tokenEstimate !== "number" ||
    !Array.isArray(value.selectedViews) ||
    !Array.isArray(value.selectedFiles) ||
    !Array.isArray(value.sourceAnchors) ||
    !Array.isArray(value.warnings) ||
    !Array.isArray(value.missing)
  ) {
    throw new Error(`${command} did not return a context resolution`);
  }
}

function assertKnowledgeValidateContextRunResponse(
  command: string,
  value: unknown,
): asserts value is KnowledgeValidateContextRunResponse {
  if (
    !isRecord(value) ||
    typeof value.valid !== "boolean" ||
    !Array.isArray(value.errors) ||
    !Array.isArray(value.warnings)
  ) {
    throw new Error(`${command} did not return a context run validation`);
  }
}

function requireAppServerString(value: unknown, message: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(message);
  }

  return value.trim();
}

function normalizeKnowledgeListPacksResponse(
  response: AppServerKnowledgeListPacksResponse | null | undefined,
): KnowledgeListPacksResponse {
  if (!response || typeof response !== "object") {
    throw new Error("App Server knowledgePack/list did not return packs");
  }

  if (!Array.isArray(response.packs)) {
    throw new Error("App Server knowledgePack/list did not return packs");
  }

  return {
    workingDir: requireAppServerString(
      response.workingDir,
      "App Server knowledgePack/list did not return workingDir",
    ),
    rootPath: requireAppServerString(
      response.rootPath,
      "App Server knowledgePack/list did not return rootPath",
    ),
    packs: response.packs as KnowledgePackSummary[],
  };
}

function normalizeKnowledgeReadPackResponse(
  response: AppServerKnowledgeReadPackResponse | null | undefined,
): KnowledgePackDetail {
  if (!response || typeof response !== "object") {
    throw new Error("App Server knowledgePack/read did not return pack");
  }

  assertKnowledgePackDetail(METHOD_KNOWLEDGE_PACK_READ, response.pack);
  return response.pack;
}

function normalizeKnowledgeImportSourceResponse(
  response: AppServerKnowledgeImportSourceResponse | null | undefined,
): KnowledgeImportSourceResponse {
  assertKnowledgeImportSourceResponse(METHOD_KNOWLEDGE_SOURCE_IMPORT, response);
  return response;
}

function normalizeKnowledgeCompilePackResponse(
  response: AppServerKnowledgeCompilePackResponse | null | undefined,
): KnowledgeCompilePackResponse {
  assertKnowledgeCompilePackResponse(METHOD_KNOWLEDGE_PACK_COMPILE, response);
  return response;
}

function normalizeKnowledgeSetDefaultPackResponse(
  response: AppServerKnowledgeSetDefaultPackResponse | null | undefined,
): KnowledgeSetDefaultPackResponse {
  assertKnowledgeSetDefaultPackResponse(
    METHOD_KNOWLEDGE_PACK_DEFAULT_SET,
    response,
  );
  return response;
}

function normalizeKnowledgeUpdatePackStatusResponse(
  response: AppServerKnowledgeUpdatePackStatusResponse | null | undefined,
): KnowledgeUpdatePackStatusResponse {
  assertKnowledgeUpdatePackStatusResponse(
    METHOD_KNOWLEDGE_PACK_STATUS_UPDATE,
    response,
  );
  return response;
}

function normalizeKnowledgeContextResolutionResponse(
  response: AppServerKnowledgeContextResolutionResponse | null | undefined,
): KnowledgeContextResolution {
  assertKnowledgeContextResolution(METHOD_KNOWLEDGE_CONTEXT_RESOLVE, response);
  return response as KnowledgeContextResolution;
}

function normalizeKnowledgeValidateContextRunResponse(
  response: AppServerKnowledgeValidateContextRunResponse | null | undefined,
): KnowledgeValidateContextRunResponse {
  assertKnowledgeValidateContextRunResponse(
    METHOD_KNOWLEDGE_CONTEXT_RUN_VALIDATE,
    response,
  );
  return response;
}

export type KnowledgePackStatus =
  | "draft"
  | "ready"
  | "needs-review"
  | "stale"
  | "disputed"
  | "archived"
  | string;

export interface KnowledgePackMetadata {
  name: string;
  description: string;
  type: string;
  profile?: "document-first" | "wiki-first" | "hybrid" | string | null;
  status: KnowledgePackStatus;
  version?: string | null;
  language?: string | null;
  license?: string | null;
  maintainers: string[];
  scope?: string | null;
  trust?: string | null;
  grounding?: string | null;
  runtime?: {
    mode?: "data" | "persona" | string | null;
    [key: string]: unknown;
  } | null;
  metadata?: {
    limeTemplate?: string;
    primaryDocument?: string;
    producedBy?: {
      kind?: string;
      name?: string;
      version?: string;
      digest?: string;
      [key: string]: unknown;
    };
    [key: string]: unknown;
  } | null;
}

export interface KnowledgePackSummary {
  metadata: KnowledgePackMetadata;
  rootPath: string;
  knowledgePath: string;
  defaultForWorkspace: boolean;
  updatedAt: number;
  sourceCount: number;
  documentCount?: number;
  wikiCount: number;
  compiledCount: number;
  runCount: number;
  preview?: string | null;
}

export interface KnowledgePackFileEntry {
  relativePath: string;
  absolutePath: string;
  bytes: number;
  updatedAt: number;
  sha256?: string | null;
  preview?: string | null;
}

export interface KnowledgePackDetail extends KnowledgePackSummary {
  guide: string;
  documents?: KnowledgePackFileEntry[];
  sources: KnowledgePackFileEntry[];
  wiki: KnowledgePackFileEntry[];
  compiled: KnowledgePackFileEntry[];
  runs: KnowledgePackFileEntry[];
}

export interface KnowledgeListPacksRequest {
  workingDir: string;
  includeArchived?: boolean;
}

export interface KnowledgeListPacksResponse {
  workingDir: string;
  rootPath: string;
  packs: KnowledgePackSummary[];
}

export interface KnowledgeImportSourceRequest {
  workingDir: string;
  packName: string;
  description?: string;
  packType?: string;
  language?: string;
  sourceFileName?: string;
  sourceText?: string;
}

export interface KnowledgeImportSourceResponse {
  pack: KnowledgePackDetail;
  source: KnowledgePackFileEntry;
}

export interface KnowledgeBuilderRuntimeOptions {
  enabled?: boolean;
  providerOverride?: string;
  modelOverride?: string;
  sessionId?: string;
}

export interface KnowledgeCompilePackResponse {
  pack: KnowledgePackDetail;
  selectedSourceCount: number;
  compiledView: KnowledgePackFileEntry;
  run: KnowledgePackFileEntry;
  warnings: string[];
}

export interface KnowledgeSetDefaultPackResponse {
  defaultPackName: string;
  defaultMarkerPath: string;
}

export interface KnowledgeUpdatePackStatusRequest {
  workingDir: string;
  name: string;
  status: KnowledgePackStatus;
}

export interface KnowledgeUpdatePackStatusResponse {
  pack: KnowledgePackDetail;
  previousStatus: KnowledgePackStatus;
  clearedDefault: boolean;
}

export interface KnowledgeResolveContextRequest {
  workingDir: string;
  name: string;
  packs?: Array<{
    name: string;
    activation?: "explicit" | "implicit" | "resolver-driven";
  }>;
  task?: string;
  maxChars?: number;
  activation?: "explicit" | "implicit" | "resolver-driven";
  writeRun?: boolean;
  runReason?: string;
}

export interface KnowledgeContextView {
  packName?: string | null;
  relativePath: string;
  tokenEstimate: number;
  charCount: number;
  sourceAnchors: string[];
}

export interface KnowledgeContextWarning {
  severity: "info" | "warning" | "error" | string;
  path?: string | null;
  message: string;
}

export interface KnowledgeContextResolution {
  packName: string;
  status: KnowledgePackStatus;
  grounding?: string | null;
  selectedViews: KnowledgeContextView[];
  selectedFiles: string[];
  sourceAnchors: string[];
  warnings: KnowledgeContextWarning[];
  missing: string[];
  tokenEstimate: number;
  fencedContext: string;
  runId?: string | null;
  runPath?: string | null;
}

export interface KnowledgeValidateContextRunRequest {
  workingDir: string;
  name: string;
  runPath: string;
}

export interface KnowledgeValidateContextRunResponse {
  valid: boolean;
  runId?: string | null;
  status?: string | null;
  errors: string[];
  warnings: string[];
}

export async function listKnowledgePacks(
  request: KnowledgeListPacksRequest,
  options: KnowledgeListPacksOptions = {},
): Promise<KnowledgeListPacksResponse> {
  const normalizedWorkingDir = request.workingDir.trim();
  if (!normalizedWorkingDir) {
    throw new Error(
      "workingDir is required to list App Server knowledge packs",
    );
  }

  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeListPacksResponse>(
      METHOD_KNOWLEDGE_PACK_LIST,
      {
        workingDir: normalizedWorkingDir,
        includeArchived: request.includeArchived ?? false,
      },
      options.appServerClient,
    );
  return normalizeKnowledgeListPacksResponse(response);
}

export async function getKnowledgePack(
  workingDir: string,
  name: string,
): Promise<KnowledgePackDetail> {
  const normalizedWorkingDir = workingDir.trim();
  const normalizedName = name.trim();
  if (!normalizedWorkingDir) {
    throw new Error(
      "workingDir is required to read App Server knowledge pack",
    );
  }
  if (!normalizedName) {
    throw new Error("name is required to read App Server knowledge pack");
  }

  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeReadPackResponse>(
      METHOD_KNOWLEDGE_PACK_READ,
      {
        workingDir: normalizedWorkingDir,
        name: normalizedName,
      },
    );
  return normalizeKnowledgeReadPackResponse(response);
}

export async function importKnowledgeSource(
  request: KnowledgeImportSourceRequest,
): Promise<KnowledgeImportSourceResponse> {
  const normalizedWorkingDir = requireAppServerString(
    request.workingDir,
    "workingDir is required to import App Server knowledge source",
  );
  const normalizedPackName = requireAppServerString(
    request.packName,
    "packName is required to import App Server knowledge source",
  );
  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeImportSourceResponse>(
      METHOD_KNOWLEDGE_SOURCE_IMPORT,
      {
        ...request,
        workingDir: normalizedWorkingDir,
        packName: normalizedPackName,
      },
    );
  return normalizeKnowledgeImportSourceResponse(response);
}

export async function compileKnowledgePack(
  workingDir: string,
  name: string,
  builderRuntime?: KnowledgeBuilderRuntimeOptions,
): Promise<KnowledgeCompilePackResponse> {
  const normalizedWorkingDir = requireAppServerString(
    workingDir,
    "workingDir is required to compile App Server knowledge pack",
  );
  const normalizedName = requireAppServerString(
    name,
    "name is required to compile App Server knowledge pack",
  );
  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeCompilePackResponse>(
      METHOD_KNOWLEDGE_PACK_COMPILE,
      {
        workingDir: normalizedWorkingDir,
        name: normalizedName,
        ...(builderRuntime ? { builderRuntime } : {}),
      },
    );
  return normalizeKnowledgeCompilePackResponse(response);
}

export async function setDefaultKnowledgePack(
  workingDir: string,
  name: string,
): Promise<KnowledgeSetDefaultPackResponse> {
  const normalizedWorkingDir = requireAppServerString(
    workingDir,
    "workingDir is required to set App Server default knowledge pack",
  );
  const normalizedName = requireAppServerString(
    name,
    "name is required to set App Server default knowledge pack",
  );
  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeSetDefaultPackResponse>(
      METHOD_KNOWLEDGE_PACK_DEFAULT_SET,
      {
        workingDir: normalizedWorkingDir,
        name: normalizedName,
      },
    );
  return normalizeKnowledgeSetDefaultPackResponse(response);
}

export async function updateKnowledgePackStatus(
  request: KnowledgeUpdatePackStatusRequest,
): Promise<KnowledgeUpdatePackStatusResponse> {
  const normalizedWorkingDir = requireAppServerString(
    request.workingDir,
    "workingDir is required to update App Server knowledge pack status",
  );
  const normalizedName = requireAppServerString(
    request.name,
    "name is required to update App Server knowledge pack status",
  );
  const normalizedStatus = requireAppServerString(
    request.status,
    "status is required to update App Server knowledge pack status",
  );
  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeUpdatePackStatusResponse>(
      METHOD_KNOWLEDGE_PACK_STATUS_UPDATE,
      {
        ...request,
        workingDir: normalizedWorkingDir,
        name: normalizedName,
        status: normalizedStatus,
      },
    );
  return normalizeKnowledgeUpdatePackStatusResponse(response);
}

export async function resolveKnowledgeContext(
  request: KnowledgeResolveContextRequest,
): Promise<KnowledgeContextResolution> {
  const normalizedWorkingDir = requireAppServerString(
    request.workingDir,
    "workingDir is required to resolve App Server knowledge context",
  );
  const normalizedName = requireAppServerString(
    request.name,
    "name is required to resolve App Server knowledge context",
  );
  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeContextResolutionResponse>(
      METHOD_KNOWLEDGE_CONTEXT_RESOLVE,
      {
        ...request,
        workingDir: normalizedWorkingDir,
        name: normalizedName,
      },
    );
  return normalizeKnowledgeContextResolutionResponse(response);
}

export async function validateKnowledgeContextRun(
  request: KnowledgeValidateContextRunRequest,
): Promise<KnowledgeValidateContextRunResponse> {
  const normalizedWorkingDir = requireAppServerString(
    request.workingDir,
    "workingDir is required to validate App Server knowledge context run",
  );
  const normalizedName = requireAppServerString(
    request.name,
    "name is required to validate App Server knowledge context run",
  );
  const normalizedRunPath = requireAppServerString(
    request.runPath,
    "runPath is required to validate App Server knowledge context run",
  );
  const response =
    await requestKnowledgeAppServer<AppServerKnowledgeValidateContextRunResponse>(
      METHOD_KNOWLEDGE_CONTEXT_RUN_VALIDATE,
      {
        ...request,
        workingDir: normalizedWorkingDir,
        name: normalizedName,
        runPath: normalizedRunPath,
      },
    );
  return normalizeKnowledgeValidateContextRunResponse(response);
}

File: lime-rs/crates/app-server/src/local_data_source/knowledge.rs
use super::data_error;
use super::value_from_serializable;
use super::values_from_serializable_vec;
use crate::RuntimeCoreError;
use app_server_protocol::KnowledgeCompilePackResponse;
use app_server_protocol::KnowledgeContextResolutionResponse;
use app_server_protocol::KnowledgeImportSourceParams;
use app_server_protocol::KnowledgeImportSourceResponse;
use app_server_protocol::KnowledgeListPacksParams;
use app_server_protocol::KnowledgeListPacksResponse;
use app_server_protocol::KnowledgeReadPackParams;
use app_server_protocol::KnowledgeReadPackResponse;
use app_server_protocol::KnowledgeResolveContextPackParams;
use app_server_protocol::KnowledgeResolveContextParams;
use app_server_protocol::KnowledgeSetDefaultPackParams;
use app_server_protocol::KnowledgeSetDefaultPackResponse;
use app_server_protocol::KnowledgeUpdatePackStatusParams;
use app_server_protocol::KnowledgeUpdatePackStatusResponse;
use app_server_protocol::KnowledgeValidateContextRunParams;
use app_server_protocol::KnowledgeValidateContextRunResponse;

pub(crate) fn list_knowledge_packs(
    params: KnowledgeListPacksParams,
) -> Result<KnowledgeListPacksResponse, RuntimeCoreError> {
    let response =
        lime_knowledge::list_knowledge_packs(lime_knowledge::KnowledgeListPacksRequest {
            working_dir: params.working_dir,
            include_archived: params.include_archived,
        })
        .map_err(data_error)?;
    Ok(KnowledgeListPacksResponse {
        working_dir: response.working_dir,
        root_path: response.root_path,
        packs: values_from_serializable_vec(response.packs)?,
    })
}

pub(crate) fn read_knowledge_pack(
    params: KnowledgeReadPackParams,
) -> Result<KnowledgeReadPackResponse, RuntimeCoreError> {
    let pack = lime_knowledge::get_knowledge_pack(lime_knowledge::KnowledgeGetPackRequest {
        working_dir: params.working_dir,
        name: params.name,
    })
    .map_err(data_error)?;

    Ok(KnowledgeReadPackResponse {
        pack: value_from_serializable(pack)?,
    })
}

pub(crate) fn import_knowledge_source(
    params: KnowledgeImportSourceParams,
) -> Result<KnowledgeImportSourceResponse, RuntimeCoreError> {
    let response =
        lime_knowledge::import_knowledge_source(lime_knowledge::KnowledgeImportSourceRequest {
            working_dir: params.working_dir,
            pack_name: params.pack_name,
            description: params.description,
            pack_type: params.pack_type,
            language: params.language,
            source_file_name: params.source_file_name,
            source_text: params.source_text,
        })
        .map_err(data_error)?;

    Ok(KnowledgeImportSourceResponse {
        pack: value_from_serializable(response.pack)?,
        source: value_from_serializable(response.source)?,
    })
}

pub(crate) fn compile_knowledge_pack(
    request: lime_knowledge::KnowledgeCompilePackRequest,
) -> Result<KnowledgeCompilePackResponse, RuntimeCoreError> {
    let response = lime_knowledge::compile_knowledge_pack(request).map_err(data_error)?;

    Ok(KnowledgeCompilePackResponse {
        pack: value_from_serializable(response.pack)?,
        selected_source_count: response.selected_source_count,
        compiled_view: value_from_serializable(response.compiled_view)?,
        run: value_from_serializable(response.run)?,
        warnings: response.warnings,
    })
}

pub(crate) fn set_default_knowledge_pack(
    params: KnowledgeSetDefaultPackParams,
) -> Result<KnowledgeSetDefaultPackResponse, RuntimeCoreError> {
    let response = lime_knowledge::set_default_knowledge_pack(
        lime_knowledge::KnowledgeSetDefaultPackRequest {
            working_dir: params.working_dir,
            name: params.name,
        },
    )
    .map_err(data_error)?;

    Ok(KnowledgeSetDefaultPackResponse {
        default_pack_name: response.default_pack_name,
        default_marker_path: response.default_marker_path,
    })
}

pub(crate) fn update_knowledge_pack_status(
    params: KnowledgeUpdatePackStatusParams,
) -> Result<KnowledgeUpdatePackStatusResponse, RuntimeCoreError> {
    let response = lime_knowledge::update_knowledge_pack_status(
        lime_knowledge::KnowledgeUpdatePackStatusRequest {
            working_dir: params.working_dir,
            name: params.name,
            status: params.status,
        },
    )
    .map_err(data_error)?;

    Ok(KnowledgeUpdatePackStatusResponse {
        pack: value_from_serializable(response.pack)?,
        previous_status: response.previous_status,
        cleared_default: response.cleared_default,
    })
}

pub(crate) fn resolve_knowledge_context(
    params: KnowledgeResolveContextParams,
) -> Result<KnowledgeContextResolutionResponse, RuntimeCoreError> {
    let response =
        lime_knowledge::resolve_knowledge_context(lime_knowledge::KnowledgeResolveContextRequest {
            working_dir: params.working_dir,
            name: params.name,
            packs: params
                .packs
                .into_iter()
                .map(to_lime_knowledge_context_pack_request)
                .collect(),
            task: params.task,
            max_chars: params.max_chars,
            activation: params.activation,
            write_run: params.write_run,
            run_reason: params.run_reason,
        })
        .map_err(data_error)?;

    Ok(KnowledgeContextResolutionResponse {
        pack_name: response.pack_name,
        status: response.status,
        grounding: response.grounding,
        selected_views: values_from_serializable_vec(response.selected_views)?,
        selected_files: response.selected_files,
        source_anchors: response.source_anchors,
        warnings: values_from_serializable_vec(response.warnings)?,
        missing: response.missing,
        token_estimate: response.token_estimate,
        fenced_context: response.fenced_context,
        run_id: response.run_id,
        run_path: response.run_path,
    })
}

pub(crate) fn validate_knowledge_context_run(
    params: KnowledgeValidateContextRunParams,
) -> Result<KnowledgeValidateContextRunResponse, RuntimeCoreError> {
    let response = lime_knowledge::validate_knowledge_context_run(
        lime_knowledge::KnowledgeValidateContextRunRequest {
            working_dir: params.working_dir,
            name: params.name,
            run_path: params.run_path,
        },
    )
    .map_err(data_error)?;

    Ok(KnowledgeValidateContextRunResponse {
        valid: response.valid,
        run_id: response.run_id,
        status: response.status,
        errors: response.errors,
        warnings: response.warnings,
    })
}

fn to_lime_knowledge_context_pack_request(
    params: KnowledgeResolveContextPackParams,
) -> lime_knowledge::KnowledgeResolveContextPackRequest {
    lime_knowledge::KnowledgeResolveContextPackRequest {
        name: params.name,
        activation: params.activation,
    }
}
